scripts/analyze_csv_data.py

```diff
@@ -0,0 +1,771 @@
+    dfs = []
+    for e in experiments:
+        old_df = df[df['experiment_label'] == e]
+        tol = old_df['tolerance'].unique()[0]
+        new_df = old_df.set_index('instance_name').reindex(
+            new_index).reset_index()
+        # otherwise these would be nan
+        new_df['tolerance'] = tol
+        new_df['experiment_label'] = e
+        dfs.append(new_df)
+    return pd.concat(dfs)
+
+
+def improvements_plot(dfs, prefix, key, ascending):
+    normalized_dfs = []
+    for df in dfs:
+        df[key] /= df[df['Experiment'] == 'PDHG'][key].to_numpy()[0]
+        normalized_dfs.append(df)
+
+    df = pd.concat(normalized_dfs)
+    fig = plt.figure(figsize=(10, 6))
+    for tol in df['tolerance'].unique():
+        _df = df[df['tolerance'] == tol].reset_index(drop=True)
+        plt.plot(
+            _df[key].to_numpy(),
+            linestyle='--',
+            marker='o',
+            label=f'tolerance {tol:.0E}')
+        plt.yscale('log')
+        plt.ylabel('Normalized ' + key)
+        plt.title(sanitize_title(prefix))
+        plt.xticks(range(len(_df['Experiment'])), _df['Experiment'].to_list())
+        ax = plt.gca()
+        ax.yaxis.set_major_locator(ticker.LogLocator(subs=[1, 2, 3, 5, 7]))
+        ax.yaxis.set_major_formatter(
+            ticker.LogFormatterSciNotation(
+                labelOnlyBase=False,
+                minor_thresholds=(4, 2)))
+        # ax.yaxis.set_major_formatter(ticker.FormatStrFormatter("%.2f")
+
+    if len(dfs) > 1:
+        plt.legend(loc='best')
+    name = key.replace(' ', '_')
+    path = os.path.join(FIGS_DIR, f'{prefix}_{name}.pdf')
+    plt.savefig(
+        path,
+        bbox_inches="tight")
+
+
+def gen_all_improvement_plots(outputs, prefix):
+    dfs = []
+    for tol, df in outputs.items():
+        df = df.copy()
+        df['tolerance'] = tol
+        dfs.append(df)
+    improvements_plot(
+        dfs,
+        prefix,
+        'KKT passes SGM10',
+        ascending=False)
+    improvements_plot(
+        dfs,
+        prefix,
+        'Solve time secs SGM10',
+        ascending=False)
+    improvements_plot(
+        dfs,
+        prefix,
+        'Solved count',
+        ascending=True)
+
+
+# First, make output directories
+if not os.path.exists(FIGS_DIR):
+    os.makedirs(FIGS_DIR)
+if not os.path.exists(TEX_DIR):
+    os.makedirs(TEX_DIR)
+
+# Get clean list of all problems we tested on:
+with open('../benchmarking/miplib2017_instance_list') as f:
+    miplib_instances = f.readlines()
+miplib_instances = [p.strip() for p in miplib_instances if p[0] != '#']
+
+with open('../benchmarking/mittelmann_instance_list') as f:
+    mittelmann_instances = f.readlines()
+mittelmann_instances = [p.strip() for p in mittelmann_instances if p[0] != '#']
+
+# Pull out 'default' (ie best) pdhg implementation to compare against:
+df_default = pd.read_csv(
+    os.path.join(
+        CSV_DIR,
+        'miplib_pdhg_enhanced_100k.csv'))
```

Comment:
We should probably also change all of the csv filenames to reflect the "mip_relaxations" and "lp_benchmarks" names.  But there are potential complications, since the experiment_name is also contained within the rows of the csv file.  I suggest we defer that to another PR (or don't fix).  Created Issue #80 to track.